### DISCUSSION TITLE
Concatenate strings with logger source generator

### DISCUSSION THREAD

**Discussion by mu88** [dotnet/runtime#98404] [Q&A]
Hi 👋🏻

We're making heavy use of [compile-time logging source generation](https://learn.microsoft.com/en-us/dotnet/core/extensions/logger-message-generator).

When it comes to non-trivial parameters like `List<string>`, I always end up with code like that:
```csharp
[LoggerMessage(
    EventId = 1234,
    EventName = FeaturePrefix + nameof(AvailableFruits),
    Level = LogLevel.Information,
    Message = "The fruits '{concatenatedFruitNames}' are available",
    SkipEnabledCheck = true)]
public static partial void AvailableFruits(this ILogger logger, string concatenatedFruitNames);

public static void AvailableFruits(this ILogger logger, List<string> fruitNames) => 
    logger.AvailableFruits(string.Join(",", fruitNames));
```

However, I'm not a big fan of this pattern due to the following reasons:
- More code to maintain.
- Auto-completion offers two logging methods.
- It makes testing like this impossible as the method signatures match:
```csharp
_logger.ShouldHaveLogged(() => LoggerExtensions.CollisionInOverrule);
```

I'd like to know whether it is possible to simplify the code, e. g. by hooking into the source generation process always execute `string.Join` when using a parameter `List<string>`?

Thank you 👋🏻

**Comment by ovska** [comment]
IME most loggers can format `IEnumerable<>` just fine. For example, with `AddConsole` your example prints just fine calling the List overload:
```cs
logger.AvailableFruits(["apple", "banana"]);
// output: The fruits 'apple, banana' are available
```

**Reply by mu88** [reply]
Strange, for me, it sometimes logs ``System.Collections.Generic.List`1[System.String]`` 🤔

**Reply by mu88** [reply]
@richlander maybe you can help: I'm using the following log message template:

```c#
[LoggerMessage(
    EventId = 12345,
    EventName = nameof(RegisteredProcessors),
    Level = LogLevel.Information,
    Message = "Registered processors in priority order: {ProcessorNames}.",
    SkipEnabledCheck = true)]
public static partial void RegisteredProcessors(this ILogger logger, IEnumerable<string> processorNames);
```

This gives me the following log output in Splunk:

```text
2024-06-03 11:15:12.740+00:00 level=INFO eventId=23010 eventName=Processing_RegisteredProcessors sourceContext="MyCompany.MyApp.Features.Processing.Processors.BulkProcessor" message="Registered processors in priority order: ProcessorA, ProcessorB, ProcessorC." processorNames="System.Linq.Enumerable+SelectIPartitionIterator`2[MyCompany.MyApp.Features.Processing.Processors.IEntityProcessor,System.String]" spanId="e4e4a9a8e1f1917f" traceId="868691a0f4d04cb852d7aa3872ebef79" parentId="0000000000000000"
```

So as you can see, the log message properly concatenates the strings. However, the attribute `processorNames` does not.
